### PR TITLE
fix(simulator): register the out-of-tree plugins before initializing configurations

### PR DESCRIPTION
**Pull request by LY-today**
#### What type of PR is this?
bug

#### What this PR does / why we need it:

Lack of adding user-defined plug-in configuration to defaultpls, resulting in loss of user configuration
#### Which issue(s) this PR fixes:

Fixes # Safeguard user expectations

#### Special notes for your reviewer:

/label tide/merge-method-squash

### CODE REVIEW

**Comment by LY-today** [comment]
@196Ikuchil @sanposhiho Hello, the ci link seems to be stuck.

**Comment by ditingdapeng** [comment]
I'm also experiencing this confusion, and I'm glad to see someone bringing it up. I would appreciate it if the author could take a look at this issue.

**Comment by Park-Jiyeonn** [comment]
Great work on adding user-defined plugin configs! This will definitely enhance customization for users. Looking forward to seeing it merged

**Comment by sanposhiho** [comment]
Did you followed the doc to integrate your plugin in the simulator? This change shouldn't be needed if you did.
https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/debuggable-scheduler.md#integrate-your-plugins-to-the-simulator

**Comment by LY-today** [comment]
> 您是否按照文档将插件集成到模拟器中？如果按照说明进行操作，则不需要进行此更改。https [://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/debuggable-scheduler.md#integrate-your-plugins-to-the-simulator](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/docs/debuggable-scheduler.md#integrate-your-plugins-to-the-simulator)

@sanposhiho I configured according to the document, and I have successfully integrated and run my own plugin on the simulator. Her name is Sample. I also understand the internal implementation principles of the simulator, but when I configure the pluginConfig for Sample, the internal implementation of the simulator KubeSchedulerConfiguration did not create SampleWrapped pluginConfig for me, causing my Sample plugin to be unable to obtain parameter information normally when executed.

**Comment by LY-today** [comment]
The external configuration is as follows：
`
apiVersion: kubescheduler.config.k8s.io/v1
kind: KubeSchedulerConfiguration
profiles:
  - schedulerName: default-scheduler
    pluginConfig:
      - args:
          prometheus_endpoint: prometheus.lens-metrics
          max_memory: 6
        name: Sample
    plugins:
      filter:
        enabled:
          - name: "Sample"
      preBind:
        enabled:
          - name: "Sample"
      postFilter:
        enabled:
          - name: "Sample"
      score:
        enabled:
          - name: "Sample"
`

The configuration generated by the simulator is as follows：
`
{
  "parallelism": 16,
  "leaderElection": {
    "leaderElect": true,
    "leaseDuration": "15s",
    "renewDeadline": "10s",
    "retryPeriod": "2s",
    "resourceLock": "leases",
    "resourceName": "kube-scheduler",
    "resourceNamespace": "kube-system"
  },
  "clientConnection": {
    "kubeconfig": "",
    "acceptContentTypes": "",
    "contentType": "application/vnd.kubernetes.protobuf",
    "qps": 50,
    "burst": 100
  },
  "enableProfiling": true,
  "enableContentionProfiling": true,
  "percentageOfNodesToScore": 0,
  "podInitialBackoffSeconds": 1,
  "podMaxBackoffSeconds": 10,
  "profiles": [
    {
      "schedulerName": "default-scheduler",
      "plugins": {
        "preEnqueue": {

        },
        "queueSort": {

        },
        "preFilter": {

        },
        "filter": {
          "enabled": [
            {
              "name": "SampleWrapped",
              "weight": 0
            }
          ]
        },
        "postFilter": {
          "enabled": [
            {
              "name": "SampleWrapped",
              "weight": 0
            }
          ]
        },
        "preScore": {

        },
        "score": {
          "enabled": [
            {
              "name": "SampleWrapped",
              "weight": 0
            }
          ]
        },
        "reserve": {

        },
        "permit": {

        },
        "preBind": {
          "enabled": [
            {
              "name": "SampleWrapped",
              "weight": 0
            }
          ]
        },
        "bind": {

        },
        "postBind": {

        },
        "multiPoint": {
          "enabled": [
            {
              "name": "SchedulingGatesWrapped",
              "weight": 0
            },
            {
              "name": "PrioritySortWrapped",
              "weight": 0
            },
            {
              "name": "NodeUnschedulableWrapped",
              "weight": 0
            },
            {
              "name": "NodeNameWrapped",
              "weight": 0
            },
            {
              "name": "TaintTolerationWrapped",
              "weight": 3
            },
            {
              "name": "NodeAffinityWrapped",
              "weight": 2
            },
            {
              "name": "NodePortsWrapped",
              "weight": 0
            },
            {
              "name": "NodeResourcesFitWrapped",
              "weight": 1
            },
            {
              "name": "VolumeRestrictionsWrapped",
              "weight": 0
            },
            {
              "name": "EBSLimitsWrapped",
              "weight": 0
            },
            {
              "name": "GCEPDLimitsWrapped",
              "weight": 0
            },
            {
              "name": "NodeVolumeLimitsWrapped",
              "weight": 0
            },
            {
              "name": "AzureDiskLimitsWrapped",
              "weight": 0
            },
            {
              "name": "VolumeBindingWrapped",
              "weight": 0
            },
            {
              "name": "VolumeZoneWrapped",
              "weight": 0
            },
            {
              "name": "PodTopologySpreadWrapped",
              "weight": 2
            },
            {
              "name": "InterPodAffinityWrapped",
              "weight": 2
            },
            {
              "name": "DefaultPreemptionWrapped",
              "weight": 0
            },
            {
              "name": "NodeResourcesBalancedAllocationWrapped",
              "weight": 1
            },
            {
              "name": "ImageLocalityWrapped",
              "weight": 1
            },
            {
              "name": "DefaultBinderWrapped",
              "weight": 0
            }
          ],
          "disabled": [
            {
              "name": "*"
            }
          ]
        }
      },
      "pluginConfig": [
        {
          "name": "VolumeBinding",
          "args": {
            "kind": "VolumeBindingArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "bindTimeoutSeconds": 600
          }
        },
        {
          "name": "Sample",
          "args": {
            "max_memory": 6,
            "prometheus_endpoint": "prometheus.lens-metrics"
          }
        },
        {
          "name": "DefaultPreemption",
          "args": {
            "kind": "DefaultPreemptionArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "minCandidateNodesPercentage": 10,
            "minCandidateNodesAbsolute": 100
          }
        },
        {
          "name": "InterPodAffinity",
          "args": {
            "kind": "InterPodAffinityArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "hardPodAffinityWeight": 1,
            "ignorePreferredTermsOfExistingPods": false
          }
        },
        {
          "name": "NodeAffinity",
          "args": {
            "kind": "NodeAffinityArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1"
          }
        },
        {
          "name": "NodeResourcesBalancedAllocation",
          "args": {
            "kind": "NodeResourcesBalancedAllocationArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "resources": [
              {
                "name": "cpu",
                "weight": 1
              },
              {
                "name": "memory",
                "weight": 1
              }
            ]
          }
        },
        {
          "name": "NodeResourcesFit",
          "args": {
            "kind": "NodeResourcesFitArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "scoringStrategy": {
              "type": "LeastAllocated",
              "resources": [
                {
                  "name": "cpu",
                  "weight": 1
                },
                {
                  "name": "memory",
                  "weight": 1
                }
              ]
            }
          }
        },
        {
          "name": "PodTopologySpread",
          "args": {
            "kind": "PodTopologySpreadArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "defaultingType": "System"
          }
        },
        {
          "name": "NodeAffinityWrapped",
          "args": {
            "kind": "NodeAffinityArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1"
          }
        },
        {
          "name": "NodeResourcesFitWrapped",
          "args": {
            "kind": "NodeResourcesFitArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "scoringStrategy": {
              "type": "LeastAllocated",
              "resources": [
                {
                  "name": "cpu",
                  "weight": 1
                },
                {
                  "name": "memory",
                  "weight": 1
                }
              ]
            }
          }
        },
        {
          "name": "VolumeBindingWrapped",
          "args": {
            "kind": "VolumeBindingArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "bindTimeoutSeconds": 600
          }
        },
        {
          "name": "PodTopologySpreadWrapped",
          "args": {
            "kind": "PodTopologySpreadArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "defaultingType": "System"
          }
        },
        {
          "name": "InterPodAffinityWrapped",
          "args": {
            "kind": "InterPodAffinityArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "hardPodAffinityWeight": 1,
            "ignorePreferredTermsOfExistingPods": false
          }
        },
        {
          "name": "DefaultPreemptionWrapped",
          "args": {
            "kind": "DefaultPreemptionArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "minCandidateNodesPercentage": 10,
            "minCandidateNodesAbsolute": 100
          }
        },
        {
          "name": "NodeResourcesBalancedAllocationWrapped",
          "args": {
            "kind": "NodeResourcesBalancedAllocationArgs",
            "apiVersion": "kubescheduler.config.k8s.io/v1",
            "resources": [
              {
                "name": "cpu",
                "weight": 1
              },
              {
                "name": "memory",
                "weight": 1
              }
            ]
          }
        }
      ]
    }
  ]
}
`

The missing configuration is as follows：
`
{
          "name": "SampleWrapped",
          "args": {
            "max_memory": 6,
            "prometheus_endpoint": "prometheus.lens-metrics"
          }
        }
`

**Comment by LY-today** [comment]
@sanposhiho 
Additional details：
```
func main() {
	command, cancelFn, err := debuggablescheduler.NewSchedulerCommand(
		debuggablescheduler.WithPlugin(sample.Name, sample.New1),
	)
	if err != nil {
		klog.Info(fmt.Sprintf("failed to build the debuggablescheduler command: %+v", err))
		os.Exit(1)
	}
	code := cli.Run(command)

	cancelFn()
	os.Exit(code)
}
```

**Comment by LY-today** [comment]
Explain that there are no errors in the process
image

**Comment by LY-today** [comment]
However, because the simulator lacks the pluginConfig for rendering SampleWrapped, the strategy cannot obtain parameters normally. sampleArgs is nil.It wasn't until my MR was adjusted that it could obtain parameter information normally.
image

**Comment by sanposhiho** [comment]
Can you help debugging/investigating the cause a bit more?

Our expectation is
1. `debuggablescheduler.WithPlugin` option eventually registers your plugin as [out of tree plugins](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/scheduler/config/plugin.go#L10-L12) ([ref](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/pkg/debuggablescheduler/debuggable_scheduler.go#L131)).
2. `RegisteredMultiPointPluginNames` [here](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/scheduler/plugin/plugins.go#L150-L153) returns your plugin name as well. 

If it works intentionally, we shouldn't have to re-register custom plugin names in `defaultpls`, like this PR tries to change.

**Comment by LY-today** [comment]
> Can you help debugging/investigating the cause a bit more?
> 
> Our expectation is
> 
> 1. `debuggablescheduler.WithPlugin` option eventually registers your plugin as [out of tree plugins](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/scheduler/config/plugin.go#L10-L12) ([ref](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/pkg/debuggablescheduler/debuggable_scheduler.go#L131)).
> 2. `RegisteredMultiPointPluginNames` [here](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/scheduler/plugin/plugins.go#L150-L153) returns your plugin name as well.
> 
> If it works intentionally, we shouldn't have to re-register custom plugin names in `defaultpls`, like this PR tries to change.

@sanposhiho Of course there is no problem. The reality is that I have debugged the entire code of the simulator. From your description, it seems that you did not understand the meaning of my PR. What I want to express is: the Sample plugin I extended can be executed by the simulator normally, but because the simulator did not encapsulate my plugin config parameters, the Sample plugin was It was executed, but the execution logic was meaningless because of missing parameters.

**Comment by sanposhiho** [comment]
> the simulator did not encapsulate my plugin config parameters

I meant, the simulator did not encapsulate your plugin config parameters because `defaultpls` didn't have your plugin name, right? So, this PR tries to put your plugin name (from a given config) to `defaultpls`.
But, our expectation is `defaultpls` should have your plugin name if you use `debuggablescheduler.WithPlugin()` option, which looks like doesn't work intentionally at some point on its way.

**Comment by LY-today** [comment]
> > the simulator did not encapsulate my plugin config parameters
> 
> I meant, the simulator did not encapsulate your plugin config parameters because `defaultpls` didn't have your plugin name, right? So, this PR tries to put your plugin name (from a given config) to `defaultpls`. But, our expectation is `defaultpls` should have your plugin name if you use `debuggablescheduler.WithPlugin()` option, which looks like doesn't work intentionally at some point on its way.

@sanposhiho If this is intentional, then I would like to ask you, under what circumstances is the user's pluginConfigs not encapsulated? I think if this was implemented in this way from the beginning of the functional design, then you should have understood it in advance.

**Comment by LY-today** [comment]
@sanposhiho If it wasn't decided during design, it would look more like a bug.

**Comment by sanposhiho** [comment]
As I said `which looks like doesn't work intentionally at some point on its way`, it's a bug, Yes. And I [asked](https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/397#issuecomment-2506328673) you to help debugging more to see where this expectation doesn't work expectedly.

**Comment by LY-today** [comment]
@sanposhiho After thinking about it, I know the root of the problem. outOfTreeRegistries will record the policy information of user-defined extensions, but the recording time is later than the execution of NewPluginConfig. As a result, when NewPluginConfig calls RegisteredMultiPointPluginNames, the user-defined extensions are missing from defaultpls. Naturally, the user-defined extensions are missing in the subsequent KSC. The adjustment of policy parameters does not involve the encapsulation of custom extensions.

**Comment by LY-today** [comment]
@sanposhiho SetOutOfTreeRegistries calls RegisteredMultiPointPluginNames later than NewPluginConfig, which is why the extended Sample plugin name cannot be obtained.

**Comment by sanposhiho** [comment]
> NewPluginConfig calls RegisteredMultiPointPluginNames earlier than SetOutOfTreeRegistries, which is why the extended Sample plugin name cannot be obtained.

Bingo!
Can we solve the problem by moving [these lines](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/pkg/debuggablescheduler/command.go#L30-L33) before [NewConfig](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/pkg/debuggablescheduler/command.go#L18) then?

**Comment by LY-today** [comment]
> > NewPluginConfig calls RegisteredMultiPointPluginNames earlier than SetOutOfTreeRegistries, which is why the extended Sample plugin name cannot be obtained.
> 
> Bingo! Can we solve the problem by moving [these lines](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/pkg/debuggablescheduler/command.go#L30-L33) before [NewConfig](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/simulator/pkg/debuggablescheduler/command.go#L18) then?

Haha, to be honest, I thought about it at first, but I thought there would be a lot of parameters that needed to be passed. In order to solve the problem with minimal changes, I used this MR solution.

**Comment by sanposhiho** [comment]
Yes, that's also fine

**Comment by LY-today** [comment]
> Yes, that's also fine

Okay, I will provide a new commit tomorrow. Please review it after you see it.

**Comment by LY-today** [comment]
> > Yes, that's also fine
> 
> Okay, I will provide a new commit tomorrow. Please review it after you see it.

@sanposhiho Submitted, please review

**Comment by LY-today** [comment]
@sanposhiho Hello, can this be sped up? It has already affected our production environment

File: simulator/pkg/debuggablescheduler/command.go [hunk 1]
```diff
@@ -15,6 +16,11 @@ func NewSchedulerCommand(opts ...Option) (*cobra.Command, func(), error) {
 	for _, o := range opts {
 		o(opt)
 	}
+
```
Comment (sanposhiho) [inline simulator/pkg/debuggablescheduler/command.go:19]: simulatorschedulerconfig.SetOutOfTreeRegistries in CreateOptionsForPlugins func no longer needed, right?.

Comment (LY-today) [reply simulator/pkg/debuggablescheduler/command.go:19]: > simulatorschedulerconfig.SetOutOfTreeRegistries in CreateOptionsForPlugins func no longer needed, right?.

@sanposhiho Yes do I need to delete it?

Comment (LY-today) [reply simulator/pkg/debuggablescheduler/command.go:19]: @sanposhiho All adjustments have been completed

Comment (LY-today) [reply simulator/pkg/debuggablescheduler/command.go:19]: @sanposhiho Are there any plans to promote this MR today?

**Comment by LY-today** [comment]
@sanposhiho What does this require me to do?
image

**Comment by sanposhiho** [comment]
You have to format the code with:

```
make format
```

**Comment by LY-today** [comment]
> You have to format the code with:
> 
> ```
> make format
> ```

make format Error:
```
cd simulator/ && make format
find . -name "*.go" | xargs gci write --section Standard --section Default --section "Prefix(sigs.k8s.io/kube-scheduler-simulator/simulator)"
xargs: gci: No such file or directory
make[1]: *** [format] Error 127
make: *** [format] Error 2
```

**Comment by sanposhiho** [comment]
Please check out the contribution doc. You also need `make tools`
https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/CONTRIBUTING.md#getting-started

**Comment by sanposhiho** [comment]
/retitle register the out-of-tree plugins before initializing configurations

**Comment by sanposhiho** [comment]
/retitle fix(simulator): register the out-of-tree plugins before initializing configurations

**Comment by LY-today** [comment]
@sanposhiho Hello, it has been formatted

**Comment by LY-today** [comment]
@sanposhiho @196Ikuchil Thank you both for doing the final review

**Review by sanposhiho — APPROVED** [review]
/lgtm
/approve

**Pull request merged**